src/bonds.jl

```diff
@@ -11,42 +25,55 @@ struct CylindricalBondEnvelope{T} <: BondEnvelope{T}
    p0::Int
    pr::Int
    pz::Int
+   floppy::Bool
+   λ::T
 end
 
-CylindricalBondEnvelope(r0cut, rcut, zcut; p0 = 2, pr = 2, pz = 2) = 
-      CylindricalBondEnvelope(r0cut, rcut, zcut, p0, pr, pz)
-
-# we can't call this `evaluate` since that would return a Vector 
-# whereas _inner_evaluate(...) should just give us a value. 
-function _inner_evaluate(env::BondEnvelope, X::AbstractState)
-   if X.be == :env 
-      return _evaluate_env(env, X)
-   elseif X.be == :bond 
-      return _evaluate_bond(env, X)
-   else
-      error("invalid X.be value")
-   end
+CylindricalBondEnvelope(r0cut, rcut, zcut; p0 = 2, pr = 2, pz = 2, floppy = true, λ = .5) = 
+      CylindricalBondEnvelope(r0cut, rcut, zcut, p0, pr, pz, floppy, λ)
+      
+struct ElipsoidBondEnvelope{T} <: BondEnvelope{T}
+   r0cut::T
+   rcut::T
+   zcut::T  # must satisfy zcut >= r0cut/2
+   p0::Int
+   pr::Int
+   floppy::Bool
+   λ::T
 end
 
+ElipsoidBondEnvelope(r0cut, zcut, rcut; p0=2, pr=2, floppy=false, λ=.5) = ElipsoidBondEnvelope(r0cut, zcut, rcut, p0, pr, floppy, λ)
+ElipsoidBondEnvelope(r0cut, cut; p0=2, pr=2, floppy=false, λ=.5) = ElipsoidBondEnvelope(r0cut, cut, cut, p0, pr, floppy, λ)
 
-Base.filter(env::CylindricalBondEnvelope, X::AbstractState) = 
-         _inner_evaluate(env,X) != 0
-
-
-function _evaluate_bond(env::CylindricalBondEnvelope, X::AbstractState)
+function _evaluate_bond(env::BondEnvelope, X::AbstractState)
    r = norm(X.rr0)
    return ((r/env.r0cut)^2 - 1)^(env.p0) * (r <= env.r0cut)
 end
 
-
-function _evaluate_env(env::CylindricalBondEnvelope, X::AbstractState)
+function _evaluate_env(env::BondEnvelope, X::AbstractState)
    # convert to cylindrical coordinates
-   r_centre = X.rr0/2
-   r̂b = X.rr0/norm(X.rr0)
-   z = dot(X.rr - r_centre, r̂b)
-   r = norm( (X.rr - r_centre) - z * r̂b )
+   z, r = cat2cyl(env, X)
+   zeff = eff_zcut(env, X)
    # then return the correct cutoff 
-   zcuteff = env.zcut + norm(X.rr0) / 2
-   return ( (z/zcuteff)^2 - 1 )^(env.pz) * (abs(z) <= zcuteff) * 
-            ( (r/env.rcut)^2 - 1 )^(env.pr) * (r <= env.rcut)
+   return envelope(env, z, r, zeff)
 end
+
+function cat2cyl(env::BondEnvelope, X::AbstractState)
+   x_centre = env.λ * X.rr0
+   x̃ = X.rr - x_centre
+   if norm(X.rr0) >0
+      r̂b = X.rr0/norm(X.rr0)
+      z̃ = dot(x̃, r̂b)
+      r̃ = norm( x̃ - z̃ * r̂b )
+   else
+      z̃ = 0.0
+      r̃ = norm(x̃)
+   end
+   return z̃, r̃
+end
+
+eff_zcut(env::BondEnvelope, X::AbstractState) = env.zcut + env.floppy * norm(env.λ * X.rr0)
+
+envelope(env::CylindricalBondEnvelope, z, r, zeff) = ( (z/zeff)^2 - 1 )^(env.pz) * (abs(z) <= zeff) * 
```

Comment:
please rename it as suggested above and document very briefly (in comments rather than doc string) the need for this function and where it is called from.